core/src/query.rs

```diff
@@ -1,77 +1,526 @@
-use crate::conn::MongoConnection;
-use crate::err::Result;
-use crate::stmt::MongoStatement;
-use bson::{Bson, Document};
-use mongodb::sync::Cursor;
+use crate::bson_type_info::BsonTypeInfo;
+use crate::{
+    conn::MongoConnection, err::Result, json_schema, stmt::MongoStatement, BsonType, BsonTypeName,
+    Error, Items, Schema,
+};
+use bson::{doc, Bson, Document};
+use itertools::Itertools;
+use mongodb::{options::AggregateOptions, sync::Cursor};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    time::Duration,
+};
 
 #[derive(Debug)]
 pub struct MongoQuery {
     // The cursor on the result set.
     resultset_cursor: Cursor<Document>,
-    // The result set metadata.
+    // The result set metadata, sorted alphabetically by collection and field name.
     resultset_metadata: Vec<MongoColMetadata>,
+    // The current deserialized "row".
+    current: Option<Document>,
 }
 
 impl MongoQuery {
-    // Create a new MongoStatement with StmtKind::Query on the connection currentDB.
-    // Executes a $sql aggregation with the given query and initialize the Resultset cursor.
-    // The query timeout comes from the statement attribute SQL_ATTR_QUERY_TIMEOUT. If there is a
-    // timeout, the query must finish before the timeout or an error is returned
+    // Create a new MongoQuery on the connection's current database. Execute a
+    // $sql aggregation with the given query and initialize the result set
+    // cursor. If there is a timeout, the query must finish before the timeout
+    // or an error is returned.
     pub fn execute(
-        _client: &MongoConnection,
-        _query_timeout: Option<i32>,
-        _query: &str,
+        client: &MongoConnection,
+        query_timeout: Option<i32>,
+        query: &str,
     ) -> Result<Self> {
-        unimplemented!()
+        match &client.current_db {
```

Comment:
[nit] I would use  
`let current_db = client.current_db.ok_or_else(||Error::NoDatabase)?;` so we can avoid a nesting depth that runs the entire method.